Add support for authenticating via user-default browser

ryanjduffy opened the pull request:
Primarily, this adds a new auth flow that launches the user's preferred browser to complete authentication. While the user completes the auth flow, the Replay browser polls (every 3 seconds over 2 minutes) the backend to retrieve the token. Once retrieved, the Replay browser exchanges the refresh token for an auth token and passes it to the app via `WebChannel`.

It continues to support the existing auth flow (even when the feature is enabled) so the user can use app's login button from the Replay browser as before.

* Adds `devtools.recordreplay.ext-auth` (defaults to `false`) to enable the new external auth flow
* Updates `api-server` to allow valid unauthenticated requests by omitting the `Authorization` header when unset
* Consolidates auth-related logic (`openSigninPage` and `initializeRecordingWebChannel`) to `auth.js`

![image](https://user-images.githubusercontent.com/788456/147532508-82f675f3-901f-46d8-aee9-eac6535eb052.png)

ryanjduffy committed:
Add support for authenticating via user-default browser

jasonLaster commented on devtools/startup/DevToolsStartup.jsm:
```diff
@@ -739,23 +740,23 @@ DevToolsStartup.prototype = {
+    //     const { token } = message;
+    //     saveRecordingToken(token);
+    //   });
+    // }
```
It's great to see this go away

ryanjduffy replied:
We _might_ have to bring it back in the opposite direction to inject auth into the app when running in the Replay browser but at least the browser will own it's own authentication process now.

jasonLaster commented on devtools/server/actors/replay/connection.js:
```diff
@@ -1469,6 +1476,45 @@ distributor.addObserver({
+    new Promise(async (resolve, reject) => {
+      let retries = 0;
+      while (retries < 20) {
+        const resp = await queryAPIServer(`
```
very cool

ryanjduffy committed:
centralize auth logic

Moves openSigninPage into auth so it can (eventually) be triggered
by app-side event.

Moves the WebChannel logic into auth so other modules aren't
brokering auth-related connections.

ryanjduffy committed:
undo a couple extra changes

ryanjduffy committed:
fix refresh token handler to fetch the token

ryanjduffy committed:
move validation flag clean up to a consistently executed place

previously, saveRecordingToken was called by the login webchannel process
but now it is not. hooking into the pref change callback is a safer place.

ryanjduffy committed:
fix using auth channel from handler

ryanjduffy committed:
fix key encoding

ryanjduffy committed:
pull view host from env variable

ryanjduffy committed:
fix build-test-branch script to support branches with forward slash

ryanjduffy committed:
add pref to select auth flow

ryanjduffy committed:
add telemetry

ryanjduffy committed:
undo github action change

ryanjduffy committed:
more diff cleanup

ryanjduffy committed:
extend retry to 2 min

bhackett1024 commented on devtools/server/actors/replay/auth.js:
```diff
@@ -65,4 +114,142 @@ function tokenExpiration(token) {
+    })
+  // TODO [ryanjduffy]: Add support for app login to use the browser auth flow
+  // } else if (type === "login") {
+  //   openSigninPage();
```
This commented out code is a little strange, can you remove it and maybe expand the comment above it?

bhackett1024 reviewed (approved):
Sorry for the delay, I didn't notice this review (btw if you're waiting for a review feel free to ping me in discord).

ryanjduffy committed:
Merge remote-tracking branch 'origin/webreplay-release' into ryan/browser-auth

ryanjduffy committed:
comment clean up
